src/url.js

```diff
@@ -217,9 +220,102 @@ export function getSourceOrigin(url) {
       'Unknown path prefix in url %s', url.href);
   const domainOrHttpsSignal = path[2];
   const origin = domainOrHttpsSignal == 's'
-      ? 'https://' + path[3]
-      : 'http://' + domainOrHttpsSignal;
+      ? 'https://' + decodeURIComponent(path[3])
+      : 'http://' + decodeURIComponent(domainOrHttpsSignal);
   // Sanity test that what we found looks like a domain.
   assert(origin.indexOf('.') > 0, 'Expected a . in origin %s', origin);
   return origin;
 }
+
+/**
+ * Returns the source URL of an AMP document for documents served
+ * on a proxy origin or directly.
+ * @param {string|!Location} url URL of an AMP document.
+ * @return {string}
+ */
+export function getSourceUrl(url) {
+  if (typeof url == 'string') {
+    url = parseUrl(url);
+  }
+
+  // Not a proxy URL - return the URL itself.
+  if (!isProxyOrigin(url)) {
+    return url.href;
+  }
+
+  // A proxy URL.
+  // Example path that is being matched here.
+  // https://cdn.ampproject.org/c/s/www.origin.com/foo/
+  // The /s/ is optional and signals a secure origin.
+  const path = url.pathname.split('/');
+  const prefix = path[1];
+  assert(prefix == 'c' || prefix == 'v',
+      'Unknown path prefix in url %s', url.href);
+  const domainOrHttpsSignal = path[2];
+  const origin = domainOrHttpsSignal == 's'
+      ? 'https://' + decodeURIComponent(path[3])
+      : 'http://' + decodeURIComponent(domainOrHttpsSignal);
+  // Sanity test that what we found looks like a domain.
+  assert(origin.indexOf('.') > 0, 'Expected a . in origin %s', origin);
+  path.splice(1, domainOrHttpsSignal == 's' ? 3 : 2);
+  return origin + path.join('/') + (url.search || '') + (url.hash || '');
+}
+
+
+/**
+ * Returns the proxy URL for the specified source URL.
+ * @param {string|!Location} url
+ * @return {string}
+ */
+export function getProxyUrl(url) {
+  if (typeof url == 'string') {
+    url = parseUrl(url);
+  }
+
+  // A proxy URL already - return the URL itself.
+  if (isProxyOrigin(url)) {
+    return url.href;
+  }
+
+  return 'https://cdn.ampproject.org' +
+      '/c' + (url.protocol == 'https:' ? '/s' : '') +
+      '/' + encodeURIComponent(url.host) +
+      url.pathname + (url.search || '') + (url.hash || '');
+}
+
+
+/**
+ * Returns absolute URL resolved based on the relative URL and the base.
+ * @param {string} relativeUrlString
+ * @param {string|!Location} baseUrl
+ * @return {string}
+ */
+export function resolveRelativeUrl(relativeUrlString, baseUrl) {
+  if (typeof baseUrl == 'string') {
+    baseUrl = parseUrl(baseUrl);
+  }
+  const relativeUrl = parseUrl(relativeUrlString);
+
+  // Absolute URL.
+  if (relativeUrlString.toLowerCase().indexOf(relativeUrl.protocol) == 0) {
+    return relativeUrl.href;
+  }
+
+  // Protocol-relative URL.
+  if (relativeUrlString.indexOf('//') == 0) {
```

Comment:
I can additionally support `//`. You are right, might be hard to find all possible cases here...